tensorflow/tools/xla_extract/tf_graph_to_xla_lib.cc

```diff
@@ -137,25 +137,39 @@ xla::HloModuleProto ExtractHloFromGraphDef(const GraphDef& in_graph,
     LOG(INFO) << "cluster not found, using "<<fdef.signature().name()<<" instead\n";
   }
 
-  // Need to rearrange these xla_args to match graph.
-  // features and labels should be the first nodes but sometimes aren't
-  // If args are [a, b, c, d, e, ...] and a is the features arg, then leave
-  // as is.  if d is the features arg, then reorder as [d, e, ..., a, b, c]
-  // This corresponds to a rotation with `d` as the axis.
   auto xla_args = BuildXlaArgsFromClientGraph(client_graph);
 
-  // Find the features arg (first parameter tensor)
-  auto features_iter = std::find_if(
-      xla_args.begin(), xla_args.end(),
-      [](const XlaCompiler::Argument& xarg) -> bool {
-        return (xarg.kind == XlaCompiler::Argument::kParameter &&
-                xarg.resource_kind == 0 && xarg.initialized == false);
-      });
 
-  // Rotate around it
-  std::rotate(xla_args.begin(), features_iter, xla_args.end());
+  // to make sure xla_args matches fdef
+  std::vector<XlaCompiler::Argument> new_xla_args;
+  auto fdef_ground = fdef.signature().input_arg();
+
+  for(int i = 0;i<xla_args.size();i++){
+    std::string xla_arg_name=xla_args[i].name;
+    std::replace(xla_arg_name.begin(),xla_arg_name.end(), '/','_');
+    std::replace(xla_arg_name.begin(),xla_arg_name.end(), '.','_');
+    xla_arg_name=xla_arg_name+"_0_arg";
+    std::transform(xla_arg_name.begin(), xla_arg_name.end(), xla_arg_name.begin(), ::tolower);
+
+    std:string readvar="readvariableop";
+    std::string read_idt="identity";
+    std::size_t pos = xla_arg_name.find(readvar);
+
+    if(pos!= std::string::npos){
+      xla_arg_name.replace(pos,readvar.length(),read_idt);
+    }
+
+    for(int j=0; j<fdef_ground.size();j++){
```

Comment:
For clarity, would prefer it if you would construct the arg names that you are looking for in one pass, and then filtered against the `fdef_ground` in a subsequent pass.  Also maybe suffix it with `ground_truth` or comment it because it is not clear where "ground" comes from

Reply:
I use the ordering from fdef to order the xla_args, and will need to store the temporary names, so not sure if it should be split into two passes.  added `truth` to the name.